### PR TITLE
Scheduling experiment

### PR DESCRIPTION
**Pull request by tzneal**

### CODE REVIEW

File: pkg/apis/provisioning/v1alpha5/requirements.go [hunk 1]
```diff
@@ -142,28 +143,54 @@ func (r Requirements) WellKnown() Requirements {
...
+			lhs.Operator = v1.NodeSelectorOpIn
+			lhs.Values = nil
+		}
+	} else if lhs.Operator == v1.NodeSelectorOpIn {
+		lhs.Values = functional.Intersect(lhs.Values, rhs.Values)
+	} else if lhs.Operator == v1.NodeSelectorOpNotIn {
+		lhs.Values = functional.Union(lhs.Values, rhs.Values)
+	} else {
+		lhs.Operator = v1.NodeSelectorOpIn
+		lhs.Values = nil
```
Comment (ellistarn) [inline pkg/apis/provisioning/v1alpha5/requirements.go:189]: What does nil represent?

File: pkg/apis/provisioning/v1alpha5/requirements.go [hunk 2]
```diff
@@ -88,12 +90,11 @@ var (
 type Requirements struct {
 	// Requirements are layered with Labels and applied to every node.
 	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
-	requirements map[string]sets.Set          `json:"-"`
```
Comment (ellistarn) [inline pkg/apis/provisioning/v1alpha5/requirements.go:91]: This was implemented recently for performance/cleanliness of implementation.

File: pkg/apis/provisioning/v1alpha5/requirements.go (hunk 1 above)
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/requirements.go:189]: It's used to construct an unsatisfiable requirement.

File: pkg/apis/provisioning/v1alpha5/requirements.go (hunk 2 above)
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/requirements.go:91]: I ended up not using it as I had found issues where the Requirements/requirements weren't kept in sync.  Instead I just added a method that could merge Requirements together and used that for identifying incompatibilities.

Comment (ellistarn) [reply pkg/apis/provisioning/v1alpha5/requirements.go:91]: There's some history we can discuss on this.

File: pkg/controllers/provisioning/scheduling/podaffinity.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	if term.PodAffinityTerm.LabelSelector == nil || len(term.PodAffinityTerm.LabelSelector.MatchLabels) == 0 {
+		return
+	}
+	p.preferredAnti = append(p.preferredAnti, term)
+	p.preferredAntiSelectors = append(p.preferredAntiSelectors, labels.SelectorFromSet(term.PodAffinityTerm.LabelSelector.MatchLabels))
+}
+
+// Score returns a numerical score and a boolean indicating if the pod affinity is a hard failure (e.g. a pod
+// violates an anti-affinity rule)
+func (p *podaffinity) Score(pods []*v1.Pod) (score int, hardFailure bool) {
```
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/podaffinity.go:69]: Why do we care about score? I understand the boolean case.

File: pkg/controllers/provisioning/scheduling/scheduler.go [hunk 4]
```diff
@@ -16,21 +16,26 @@ package scheduling
...
+
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
 	"github.com/aws/karpenter/pkg/metrics"
-	"github.com/aws/karpenter/pkg/utils/injection"
-	"github.com/aws/karpenter/pkg/utils/resources"
 )
 
+// this is used internally only to mark why we've failed to schedule a pod to
+// provide a better user experience
+const scheduleFailureReasonKey = "karpenter.sh/scheduleFailureReason"
```
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/scheduler.go:37]: unused?

File: pkg/controllers/provisioning/scheduling/podaffinity.go (hunk 3 above)
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/podaffinity.go:69]: So we can support preferred pod affinity weights, see `byPodPreference`

File: pkg/controllers/provisioning/scheduling/scheduler.go [hunk 5]
```diff
@@ -16,21 +16,26 @@ package scheduling
 
 import (
```
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/scheduler.go:17]: I think we need to move scheduling to be before the selection decision.

File: pkg/controllers/provisioning/scheduling/scheduler.go (hunk 4 above)
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduler.go:37]: No, it's used in `applyTopologyConstraints()` if applying those constraints results in some unsatisfiable condition.  The scheduler then knows that the pod is not schedulable and won't bother checking for node compatibility while also reporting why it was unschedulable.

File: pkg/controllers/provisioning/scheduling/topology.go [hunk 6]
```diff
@@ -18,129 +18,271 @@ import (
...
+	// tracking due to a pod anti-affinity
+	affinityDerivationPodAntiAffinity
 )
 
 type Topology struct {
-	kubeClient client.Client
+	Key                string
+	MaxSkew            int32
+	PodAffinityDerived affinityDerivation
+	AllowViolation     bool  // if true, this is from a preferred affinity
```
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topology.go:48]: Let's talk about how preferences work today.

**Pull request closed without merging**